openhands/sdk/conversation/stuck_detector.py

```diff
@@ -0,0 +1,271 @@
+from openhands.sdk.conversation.state import ConversationState
+from openhands.sdk.event import (
+    ActionEvent,
+    AgentErrorEvent,
+    CondensationSummaryEvent,
+    Event,
+    MessageEvent,
+    ObservationEvent,
+)
+from openhands.sdk.logger import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class StuckDetector:
+    """Detects when an agent is stuck in repetitive or unproductive patterns.
+
+    This detector analyzes the conversation history to identify various stuck patterns:
+    1. Repeating action-observation cycles
+    2. Repeating action-error cycles
+    3. Agent monologue (repeated messages without user input)
+    4. Repeating alternating action-observation patterns
+    5. Context window errors indicating memory issues
+    """
+
+    def __init__(self, state: ConversationState):
+        self.state = state
+
+    def is_stuck(self) -> bool:
+        """Check if the agent is currently stuck."""
+        events = list(self.state.events)
+
+        # Only look at history after the last user message
+        last_user_msg_index = next(
+            (
+                i
+                for i in reversed(range(len(events)))
+                if isinstance(events[i], MessageEvent) and events[i].source == "user"
+            ),
+            -1,  # Default to -1 if no user message found
```

Comment:
This should normally shouldn't happen, right? Maybe we should throw a warining for this and just `return`?